File: crane_plus_examples_py/crane_plus_examples_py/pick_and_place_tf.py
# Copyright 2025 RT Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import math

from crane_plus_examples_py.utils import plan_and_execute

from geometry_msgs.msg import PoseStamped

from moveit.core.robot_state import RobotState
from moveit.planning import (
    MoveItPy,
    PlanRequestParameters,
)
from moveit_msgs.msg import (
    BoundingVolume,
    Constraints,
    JointConstraint,
    OrientationConstraint,
    PositionConstraint,
)

import numpy as np

import rclpy
from rclpy.node import Node
from scipy.spatial.transform import Rotation
from shape_msgs.msg import SolidPrimitive
from tf2_ros import TransformException, TransformListener, TransformStamped
from tf2_ros.buffer import Buffer


class PickAndPlaceTf(Node):
    # グリッパの開閉角度
    GRIPPER_DEFAULT = 0.0
    GRIPPER_OPEN = math.radians(-30.0)
    GRIPPER_CLOSE = math.radians(10.0)

    # 待機姿勢の位置姿勢（x, y, z [m], roll, pitch, yaw [deg]）
    STANDBY = (0.0, 0.0, 0.3, 0.0, 0.0, 0.0)

    # 把持アプローチ時の高さおよびピッチ角
    GRASP_HEIGHT = 0.04
    GRASP_PITCH = 90.0

    # 搬送時の中間姿勢
    TRANSIT_1 = (0.12, 0.0, 0.17, 0.0, 90.0, 0.0)
    TRANSIT_2 = (0.0, -0.12, 0.17, 0.0, 90.0, -90.0)

    # 置く位置（プレース位置）とその退避姿勢
    PLACE = (0.0, -0.25, 0.05, 0.0, 90.0, -90.0)
    PLACE_RETRACT = (0.0, -0.25, 0.10, 0.0, 90.0, -90.0)

    def __init__(self):
        super().__init__('pick_and_place_tf')
        self.logger = self.get_logger()

        # TFリスナーの初期化
        self.tf_buffer = Buffer()
        self.tf_listener = TransformListener(self.tf_buffer, self)
        self.tf_past = TransformStamped()

        # MoveItPyのインスタンスを生成し、planning componentを取得
        self.crane_plus = MoveItPy(node_name='moveit_py')
        self.logger.info('MoveItPy instance created')

        # アーム・グリッパ制御用 planning component
        self.arm = self.crane_plus.get_planning_component('arm_tcp')
        self.gripper = self.crane_plus.get_planning_component('gripper')

        # ロボットモデルの取得（ジョイント目標値の設定に使用）
        self.robot_model = self.crane_plus.get_robot_model()

        # プランニングの設定（動作プランナーと速度・加速度スケール）
        self.arm_plan_params = PlanRequestParameters(self.crane_plus, 'ompl_rrtc')
        self.arm_plan_params.max_velocity_scaling_factor = 1.0  # Set 0.0 ~ 1.0
        self.arm_plan_params.max_acceleration_scaling_factor = 1.0  # Set 0.0 ~ 1.0

        self.gripper_plan_params = PlanRequestParameters(self.crane_plus, 'ompl_rrtc')
        self.gripper_plan_params.max_velocity_scaling_factor = 1.0  # Set 0.0 ~ 1.0
        self.gripper_plan_params.max_acceleration_scaling_factor = 1.0  # Set 0.0 ~ 1.0

        # SRDFに定義されている "home" の姿勢に移動
        self.arm.set_start_state_to_current_state()
        self.arm.set_goal_state(configuration_name='home')
        plan_and_execute(
            self.crane_plus, self.arm, self.logger,
            single_plan_parameters=self.arm_plan_params,
        )

        # アームの可動範囲制限を設定してから待機姿勢に移動する
        self.set_constraints()
        self.control_arm(*self.STANDBY)

        # 0.5秒ごとにon_timerを呼び出すタイマーを作成
        self.timer = self.create_timer(0.5, self.on_timer)

    def on_timer(self):
        # target_0のtf位置姿勢を取得
        try:
            tf_msg = self.tf_buffer.lookup_transform(
                'crane_plus_base', 'target_0', rclpy.time.Time()
            )
        except TransformException as ex:
            self.logger.info(f'Could not transform base_link to target: {ex}')
            return

        now_time = self.get_clock().now()
        FILTERING_TIME = rclpy.duration.Duration(seconds=2)
        STOP_TIME_THRESHOLD = rclpy.duration.Duration(seconds=3)
        DISTANCE_THRESHOLD = 0.01

        tf_time = rclpy.time.Time.from_msg(tf_msg.header.stamp)
        tf_elapsed_time = now_time - tf_time
        tf_past_time = rclpy.time.Time.from_msg(self.tf_past.header.stamp)
        tf_stop_time = now_time - tf_past_time

        # 現在時刻から2秒以内に受け取ったtfを使用
        if tf_elapsed_time > FILTERING_TIME:
            return

        tf_diff = np.linalg.norm(
            [
                self.tf_past.transform.translation.x - tf_msg.transform.translation.x,
                self.tf_past.transform.translation.y - tf_msg.transform.translation.y,
                self.tf_past.transform.translation.z - tf_msg.transform.translation.z,
            ]
        )

        # 把持対象の位置が停止していることを判定
        if tf_diff > DISTANCE_THRESHOLD:
            self.tf_past = tf_msg
            return

        # 把持対象が3秒以上停止している場合ピッキング動作開始
        if tf_stop_time < STOP_TIME_THRESHOLD:
            return

        self.picking(tf_msg.transform.translation)

    def picking(self, target_position):
        # 何かを掴んでいた時のためにハンドを開閉
        self.move_gripper_angle(self.GRIPPER_OPEN)

        x = target_position.x
        y = target_position.y
        theta_deg = math.degrees(math.atan2(y, x))

        # ターゲットの正面に向ける
        self.control_arm(0.0, 0.0, self.STANDBY[2], 0, 0, theta_deg)

        # ピック動作（掴みに行く）
        if not self.control_arm(x, y, self.GRASP_HEIGHT, 0, self.GRASP_PITCH, theta_deg):
            # アーム動作に失敗した場合は待機姿勢に戻る
            self.control_arm(*self.STANDBY)
            return

        self.move_gripper_angle(self.GRIPPER_CLOSE)

        # プレース動作（移動して置く）
        self.control_arm(*self.TRANSIT_1)
        self.control_arm(*self.TRANSIT_2)
        self.control_arm(*self.PLACE)
        self.move_gripper_angle(self.GRIPPER_OPEN)

        # 待機姿勢に戻る
        self.control_arm(*self.PLACE_RETRACT)
        self.control_arm(*self.STANDBY)
        self.move_gripper_angle(self.GRIPPER_DEFAULT)

    def move_gripper_angle(self, angle):
        # グリッパを角度[rad]を指定して開閉する
        self.gripper.set_start_state_to_current_state()
        robot_state = RobotState(self.robot_model)
        robot_state.set_joint_group_positions('gripper', [angle])
        self.gripper.set_goal_state(robot_state=robot_state)
        plan_and_execute(
            self.crane_plus, self.gripper, self.logger,
            single_plan_parameters=self.gripper_plan_params,
        )

    def control_arm(self, x, y, z, roll, pitch, yaw):
        # アームを目標位置（x, y, z [m]）・姿勢（roll, pitch, yaw [deg]）に動かす
        # （IKの精度を向上させるため、位置・姿勢の許容範囲を設定）
        POSITION_TOLERANCE = 0.00001
        ORIENTATION_TOLERANCE = 0.0001

        target_pose = PoseStamped()
        target_pose.header.frame_id = 'crane_plus_base'
        target_pose.pose.position.x = x
        target_pose.pose.position.y = y
        target_pose.pose.position.z = z
        rotation = Rotation.from_euler('xyz', [roll, pitch, yaw], degrees=True)
        quat = rotation.as_quat()
        target_pose.pose.orientation.x = quat[0]
        target_pose.pose.orientation.y = quat[1]
        target_pose.pose.orientation.z = quat[2]
        target_pose.pose.orientation.w = quat[3]

        goal_constraints = Constraints()
        goal_constraints.name = 'tolerance_goal'

        position_constraint = PositionConstraint()
        position_constraint.header.frame_id = 'crane_plus_base'
        position_constraint.link_name = 'crane_plus_link_tcp'
        tolerance_region = BoundingVolume()
        primitive = SolidPrimitive()
        primitive.type = SolidPrimitive.SPHERE
        primitive.dimensions = [POSITION_TOLERANCE]
        tolerance_region.primitives.append(primitive)
        tolerance_region.primitive_poses.append(target_pose.pose)
        position_constraint.constraint_region = tolerance_region
        position_constraint.weight = 1.0

        orientation_constraint = OrientationConstraint()
        orientation_constraint.header.frame_id = 'crane_plus_base'
        orientation_constraint.link_name = 'crane_plus_link_tcp'
        orientation_constraint.orientation = target_pose.pose.orientation
        orientation_constraint.absolute_x_axis_tolerance = ORIENTATION_TOLERANCE
        orientation_constraint.absolute_y_axis_tolerance = ORIENTATION_TOLERANCE
        orientation_constraint.absolute_z_axis_tolerance = ORIENTATION_TOLERANCE
        orientation_constraint.weight = 1.0

        goal_constraints.position_constraints.append(position_constraint)
        goal_constraints.orientation_constraints.append(orientation_constraint)

        self.arm.set_start_state_to_current_state()
        self.arm.set_goal_state(motion_plan_constraints=[goal_constraints])

        return plan_and_execute(
            self.crane_plus, self.arm, self.logger,
            single_plan_parameters=self.arm_plan_params,
        )

    def set_constraints(self):
        # アームの関節の一部に可動制限を設定する
        constraints = Constraints()
        constraints.name = 'arm_constraints'

        joint_constraint = JointConstraint()
        joint_constraint.joint_name = 'crane_plus_joint1'
        joint_constraint.position = 0.0
        joint_constraint.tolerance_above = math.radians(100)
        joint_constraint.tolerance_below = math.radians(100)
        joint_constraint.weight = 1.0
        constraints.joint_constraints.append(joint_constraint)

        joint_constraint = JointConstraint()
        joint_constraint.joint_name = 'crane_plus_joint3'
        joint_constraint.position = 0.0
        joint_constraint.tolerance_above = math.radians(0)
        joint_constraint.tolerance_below = math.radians(180)
        joint_constraint.weight = 1.0
        constraints.joint_constraints.append(joint_constraint)

        self.arm.set_path_constraints(constraints)

    def clear_constraints(self):
        # 設定された関節可動制限をクリアする
        self.arm.clear_path_constraints()


def main(args=None):
    rclpy.init(args=args)

    pick_and_place_tf_node = PickAndPlaceTf()

    rclpy.spin(pick_and_place_tf_node)

    pick_and_place_tf_node.clear_constraints()

    # Finish with error. Related Issue
    # https://github.com/moveit/moveit2/issues/2693
    pick_and_place_tf_node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: crane_plus_examples/src/pick_and_place_tf.cpp
// Copyright 2022 RT Corporation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Reference:
// https://github.com/ros-planning/moveit2_tutorials/blob
// /a547cf49ff7d1fe16a93dfe020c6027bcb035b51/doc/move_group_interface
// /src/move_group_interface_tutorial.cpp
// https://docs.ros.org/en/humble/Tutorials/Intermediate/Tf2/Writing-A-Tf2-Listener-Cpp.html

#include <chrono>
#include <cmath>
#include <memory>

#include "angles/angles.h"
#include "geometry_msgs/msg/pose.hpp"
#include "geometry_msgs/msg/transform_stamped.hpp"
#include "moveit/move_group_interface/move_group_interface.hpp"
#include "rclcpp/rclcpp.hpp"
#include "tf2/convert.hpp"
#include "tf2/exceptions.hpp"
#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"
#include "tf2_ros/buffer.h"
#include "tf2_ros/transform_listener.h"

using namespace std::chrono_literals;
using MoveGroupInterface = moveit::planning_interface::MoveGroupInterface;

class PickAndPlaceTf : public rclcpp::Node
{
public:
  // 位置姿勢パラメータ（x, y, z [m], roll, pitch, yaw [deg]）
  struct PoseParams
  {
    double x, y, z, roll, pitch, yaw;
  };

  // グリッパの開閉角度
  inline static const double GRIPPER_DEFAULT = 0.0;
  inline static const double GRIPPER_OPEN = angles::from_degrees(-30.0);
  inline static const double GRIPPER_CLOSE = angles::from_degrees(10.0);

  // 待機姿勢の位置姿勢（x, y, z [m], roll, pitch, yaw [deg]）
  inline static const PoseParams STANDBY = {0.0, 0.0, 0.3, 0.0, 0.0, 0.0};

  // 把持アプローチ時の高さおよびピッチ角
  inline static const double GRASP_HEIGHT = 0.04;
  inline static const double GRASP_PITCH = 90.0;

  // 搬送時の中間姿勢
  inline static const PoseParams TRANSIT_1 = {0.12, 0.0, 0.17, 0.0, 90.0, 0.0};
  inline static const PoseParams TRANSIT_2 = {0.0, -0.12, 0.17, 0.0, 90.0, -90.0};

  // 置く位置（プレース位置）とその退避姿勢
  inline static const PoseParams PLACE = {0.0, -0.25, 0.05, 0.0, 90.0, -90.0};
  inline static const PoseParams PLACE_RETRACT = {0.0, -0.25, 0.10, 0.0, 90.0, -90.0};

  PickAndPlaceTf(
    rclcpp::Node::SharedPtr move_group_arm_node,
    rclcpp::Node::SharedPtr move_group_gripper_node)
  : Node("pick_and_place_tf_node")
  {
    move_group_arm_ = std::make_shared<MoveGroupInterface>(move_group_arm_node, "arm_tcp");
    move_group_arm_->setMaxVelocityScalingFactor(1.0);
    move_group_arm_->setMaxAccelerationScalingFactor(1.0);

    move_group_gripper_ = std::make_shared<MoveGroupInterface>(move_group_gripper_node, "gripper");
    move_group_gripper_->setMaxVelocityScalingFactor(1.0);
    move_group_gripper_->setMaxAccelerationScalingFactor(1.0);

    // IKの成功率を向上させるため、目標位置姿勢の許容範囲を小さく設定
    move_group_arm_->setGoalPositionTolerance(1e-5);
    move_group_arm_->setGoalOrientationTolerance(1e-4);

    // SRDFに定義されている "home" の姿勢に移動
    move_arm_to_named_pose("home");

    // アームの可動範囲制限を設定してから待機姿勢に移動する
    set_constraints();
    control_arm(STANDBY);

    tf_buffer_ = std::make_unique<tf2_ros::Buffer>(this->get_clock());
    tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);

    // 0.5秒ごとにon_timerを呼び出すタイマーを作成
    timer_ = this->create_wall_timer(500ms, std::bind(&PickAndPlaceTf::on_timer, this));
  }

  // グリッパを角度[rad]を指定して開閉する
  void move_gripper_angle(const double angle)
  {
    auto joint_values = move_group_gripper_->getCurrentJointValues();
    joint_values[0] = angle;
    move_group_gripper_->setJointValueTarget(joint_values);
    move_group_gripper_->move();
  }

  // アームを目標位置・姿勢（Pose）に動かす
  void move_arm_to_pose(const geometry_msgs::msg::Pose & pose)
  {
    move_group_arm_->setPoseTarget(pose);
    move_group_arm_->move();
  }

  // アームを目標位置（x, y, z [m]）・姿勢（roll, pitch, yaw [deg]）に動かす
  bool control_arm(
    const double x, const double y, const double z,
    const double roll, const double pitch, const double yaw)
  {
    geometry_msgs::msg::Pose target_pose;
    tf2::Quaternion q;
    target_pose.position.x = x;
    target_pose.position.y = y;
    target_pose.position.z = z;
    q.setRPY(angles::from_degrees(roll), angles::from_degrees(pitch), angles::from_degrees(yaw));
    target_pose.orientation = tf2::toMsg(q);
    move_group_arm_->setPoseTarget(target_pose);
    moveit::core::MoveItErrorCode result = move_group_arm_->move();
    return result.val == moveit::core::MoveItErrorCode::SUCCESS;
  }

  // 位置姿勢パラメータ（PoseParams）を使ってアームを動かす
  bool control_arm(const PoseParams & pose)
  {
    return control_arm(pose.x, pose.y, pose.z, pose.roll, pose.pitch, pose.yaw);
  }

  // SRDFに定義された姿勢名でアームを動かす
  void move_arm_to_named_pose(const std::string & name)
  {
    move_group_arm_->setNamedTarget(name);
    move_group_arm_->move();
  }

  // アームの関節の一部に可動制限を設定する
  void set_constraints()
  {
    moveit_msgs::msg::Constraints constraints;
    constraints.name = "arm_constraints";

    moveit_msgs::msg::JointConstraint joint_constraint;
    joint_constraint.joint_name = "crane_plus_joint1";
    joint_constraint.position = 0.0;
    joint_constraint.tolerance_above = angles::from_degrees(100);
    joint_constraint.tolerance_below = angles::from_degrees(100);
    joint_constraint.weight = 1.0;
    constraints.joint_constraints.push_back(joint_constraint);

    joint_constraint.joint_name = "crane_plus_joint3";
    joint_constraint.position = 0.0;
    joint_constraint.tolerance_above = angles::from_degrees(0);
    joint_constraint.tolerance_below = angles::from_degrees(180);
    joint_constraint.weight = 1.0;
    constraints.joint_constraints.push_back(joint_constraint);

    move_group_arm_->setPathConstraints(constraints);
  }

  // 設定された関節可動制限をクリアする
  void clear_constraints()
  {
    move_group_arm_->clearPathConstraints();
  }

private:
  void on_timer()
  {
    // target_0（把持対象）のTFを取得
    geometry_msgs::msg::TransformStamped tf_msg;

    try {
      tf_msg = tf_buffer_->lookupTransform(
        "base_link", "target_0",
        tf2::TimePointZero);
    } catch (const tf2::TransformException & ex) {
      RCLCPP_INFO(
        this->get_logger(), "Could not transform base_link to target: %s",
        ex.what());
      return;
    }

    rclcpp::Time now = this->get_clock()->now();
    const auto FILTERING_TIME = rclcpp::Duration(2s);
    const auto STOP_TIME_THRESHOLD = rclcpp::Duration(3s);
    const double DISTANCE_THRESHOLD = 0.01;
    const double TARGET_Z_MIN_LIMIT = 0.04;

    tf2::Stamped<tf2::Transform> tf_current;
    tf2::convert(tf_msg, tf_current);

    const auto tf_elapsed_time = now - rclcpp::Time(tf_msg.header.stamp, RCL_ROS_TIME);
    const auto tf_stop_time =
      now - rclcpp::Time(tf_past_.stamp_.time_since_epoch().count(), RCL_ROS_TIME);

    // 現在時刻から2秒以内に受け取ったtfを使用
    if (tf_elapsed_time > FILTERING_TIME) {
      return;
    }

    double tf_diff = (tf_past_.getOrigin() - tf_current.getOrigin()).length();

    // 把持対象の位置が停止していることを判定
    if (tf_diff > DISTANCE_THRESHOLD) {
      tf_past_ = tf_current;
      return;
    }

    // 把持対象が3秒以上停止している場合ピッキング動作開始
    if (tf_stop_time < STOP_TIME_THRESHOLD) {
      return;
    }

    // 把持対象が低すぎる場合は把持位置を調整
    if (tf_current.getOrigin().z() < TARGET_Z_MIN_LIMIT) {
      tf_current.getOrigin().setZ(TARGET_Z_MIN_LIMIT);
    }

    picking(tf_current.getOrigin());
  }

  void picking(tf2::Vector3 target_position)
  {
    // 何かを掴んでいた時のためにハンドを開閉
    move_gripper_angle(GRIPPER_OPEN);

    double x = target_position.x();
    double y = target_position.y();
    double theta_deg = std::atan2(y, x) * 180.0 / M_PI;

    // ターゲットの正面に向ける
    control_arm(0.0, 0.0, STANDBY.z, 0, 0, theta_deg);

    // ピック動作（掴みに行く）
    if (!control_arm(x, y, GRASP_HEIGHT, 0, GRASP_PITCH, theta_deg)) {
      // アーム動作に失敗した場合は待機姿勢に戻る
      control_arm(STANDBY);
      return;
    }
    move_gripper_angle(GRIPPER_CLOSE);

    // プレース動作（移動して置く）
    control_arm(TRANSIT_1);
    control_arm(TRANSIT_2);
    control_arm(PLACE);
    move_gripper_angle(GRIPPER_OPEN);

    // 待機姿勢に戻る
    control_arm(PLACE_RETRACT);
    control_arm(STANDBY);
    move_gripper_angle(GRIPPER_DEFAULT);
  }

  std::shared_ptr<MoveGroupInterface> move_group_arm_;
  std::shared_ptr<MoveGroupInterface> move_group_gripper_;
  std::unique_ptr<tf2_ros::Buffer> tf_buffer_;
  std::shared_ptr<tf2_ros::TransformListener> tf_listener_{nullptr};
  rclcpp::TimerBase::SharedPtr timer_{nullptr};
  tf2::Stamped<tf2::Transform> tf_past_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::NodeOptions node_options;
  node_options.automatically_declare_parameters_from_overrides(true);
  auto move_group_arm_node = rclcpp::Node::make_shared("move_group_arm_node", node_options);
  auto move_group_gripper_node = rclcpp::Node::make_shared("move_group_gripper_node", node_options);

  // タイマーとTFリスナーを持つため、MultiThreadedExecutorを使用する
  rclcpp::executors::MultiThreadedExecutor exec;
  auto pick_and_place_tf_node = std::make_shared<PickAndPlaceTf>(
    move_group_arm_node,
    move_group_gripper_node);

  exec.add_node(pick_and_place_tf_node);
  exec.add_node(move_group_arm_node);
  exec.add_node(move_group_gripper_node);
  exec.spin();

  pick_and_place_tf_node->clear_constraints();
  rclcpp::shutdown();
  return 0;
}
